### ISSUE TITLE
Generate different artifacts/modules for different project components

### ISSUE THREAD

**Issue by magicDGS** [broadinstitute/gatk#3900]
The current GATK framework is increasing in size and functionality a lot (engine, CNV, different tools, etc.). For example, the gCNV code from #3838, will include a full python framework to be use in conjunction with the GATK CNV code. In gatk3, different artifacts were generated to allow custom picking of the correct dependencies. With gradle, a composite build can be done to assemble together every GATK4 sub-modules, and still being able to publish different components in different artifacts. At least I would like to have a different artifact for pure-java components separate from the rest, to be sure that python (for example) is not required.

Does some of this makes sense for you? A proposed scheme will be the following:

* `common`/`engine`: this should include the engine, utils, and everything that it can be useful by itself. This should be a dependency for every other module. Components in other sub-modules might be proposed to be moved to this one if they might be useful out of their own. If the package names does not change, the interface and usage will be unmodified, and then there is no change in the API.
* `spark`: I think that this is a nice separation from other components. In this case, this can include all code related with Spark classes for removal of the huge Spark dependency in sub-projects that does not require them.
* `tools` and `spark-tools`: this can be even split in more fine grained sub-modules depending on the pipeline (e.g. CNV, Mutect, etc., if it makes sense). In addition, the separation between normal tools and spark-tools will make easier for downstream projects to support or not spark in their code.
* `experimental`: this might contain prototype code that might change in the future, and that will be nice in terms of documentation purposes (always annotated with `@BetaFeature` or `@Experimental`, etc.). In addition, code shouldn't rely on the code in this package for anything, allowing to have experimental code for play around and remove if required, without any major version bump.
* `testing`: this will contain the testing framework. It is related with #1481 and #3567. 
* `documentation`: this might be useful for code dependent on `com.sun.javadoc` to do not interact with other classes if code for documenting a downstream project is not necessary.
* Other modules might be useful for concrete components: e.g, ., the gCNV python computational kernel implemented in #3838.

**Comment by cmnbroad** [comment]
@magicDGS I agree there would be some value in having finer grained artifacts - I would probably propose something somewhere in between what we have now and what you propose. But I suspect its not something that can happen for the initial release, which is approaching pretty quickly. Others may have a different opinion.

**Comment by magicDGS** [comment]
@cmnbroad - any advance on this? It looks like soon the uber-jar will include a lot of large files ( #4530 and https://github.com/broadinstitute/gatk/pull/4245) that are not directly related with the engine, and thus jar files for downstream projects will be huge.

I would like to have a way to select a maven artifact with the core engine (maybe including Spark) - although it will be nice also to get some classes in the tools package, but I can propose the change to the "core" package if required.

Thank you!

**Comment by lbergelson** [comment]
@magicDGS That's a good point.  It's really nasty to package those large files into the main jar.
